### PR TITLE
SettingGateway Rewrite

### PR DESCRIPTION
**Pull request by kyranet**
# SettingGateway rewrite

This PR **breaks a lot of code**, you must be careful before updating when this gets merged.

## Added
- A changelog...
- Added `KlasaMessage`, `KlasaGuild`, and `KlasaUser`, extending `DiscordJS.Message`, `DiscordJS.Guild` and `DiscordJS.User`, respectively. Many of the getters have turned properties for performance.
- Added `User#configs`, which has an empty schema by default. The developer can take advantage of this, expand the schema, and feature a full featured user configs.
- Added `configUpdateEntry`, `configDeleteEntry` and `configCreateEntry` events.
- `Configuration` class, which all the new update methods from SettingGateway, featuring full OOP and much easier to use and remember.
- `Schema` and `SchemaPiece` classes, with helper methods that allow the gateway parse and retrieve data much faster, as well as following the OOP paradigm when it comes to modifying the schema and much friendlier to use.
- More options to `GatewayDriver#add`, allowing you to set different providers in different gateway instances.
- Added **Folder** type (nested objects, yay!).
- Added `SchemaPiece.configurable` to avoid certain keys to be processed by the default conf command.
- Added **any** type to SettingsResolver. Assigning this type to a SchemaPiece will make it set the property configureable to false, it accepts anything you pass to it, even objects. So you must use dedicated commands to configure these keys instead. (For example, you can store an array of objects with a certain number of properties, something not possible in the previous versions as the conf command cannot parse objects correctly).
- **[PERF]** Much better, accurate and faster SchemaPiece validation and parsing.
- **[PERF]** Added `Schema.toJSON();`, which converts the classes into a plain JSON object. This method is 27^level times faster than iterating over it with `Object.keys`.
- Added a **Collection** provider.
- Added `Provider.cache`.
- Added GatewaySQL (extends Gateway, overriding the methods for better SQL parsing).
- Added `Gateway#insertEntry`, which inserts a new `Configuration` entry to the cache and sync if possible.

## Changed
- **[BREAKING]** Renamed `Guild#settings` and `Message#guildSettings` to `Guild#configs` and `Message#guildConfigs`, respectively.
- **[BREAKING]** Renamed `client.settings` to `client.gateways`.
- Renamed the inhibitor `requiredSettings` to `requiredConfigs`.
- Renamed the key `Command#requiredSettings` to `Command#requiredConfigs`.
- Added a new option for the `guildDelete` event: `KlasaClientOptions.preserveConfigs`, which, if true, it will not delete the guild configs when the bot leaves a guild.
- Renamed SettingGateway to Gateway.
- Renamed SettingsCache to GatewayDriver.
- **[BREAKING]** `GatewayDriver#validate` does not longer have an instance of `SettingResolver` as first parameter, but a bound `this` referring to the GatewayDriver instance. (With access to KlasaClient and SettingResolver).
- **[BREAKING]** `Gateway#get` -> `Gateway#getEntry`.
- **[BREAKING]** `Gateway#update` has been completely removed. Use `Configuration#updateOne`, `Configuration#updateArray` or `Configuration#updateMany` instead.
- **[FEATURE]** `Configuration#updateOne`, `Configuration#updateMany`, `Configuration#reset` and `Configuration#updateArray` can now filter unconfigureable keys if the parameter `avoidUnconfigurable` is passed and set to `true`.
- `Gateway#schema` is not longer a plain object, but a `Schema` instance.
- **[PERF]** Removed tuplifier in favor of cached tuples.
- **[PERF]** Improved updateArray remove method. (~1.85 times faster).
- **[PERF]** The cache now updates from the cache instead of re-syncing with the DB. (When using JSON and Collection, Samsung SSD at 600 MB/s read, performance is improved by 97k ops/sec to 53M ops/sec, 454 times faster).
- `Configuration#updateOne` now accepts array type. It'll call the private method that does the parsing for updateArray.
- **[PERF]** SQL parsing is now shared with NoSQL, however, SettingGateway's parsing procedures do not return JSON objects, instead, it returns data compatible for both environments.
- **[PERF]** When using `SettingResolver`, SettingGateway will not longer do `type.toLowerCase()`, but they're lowercase'd at startup.
- **[BREAKING]** Gateway do not longer extend SchemaManager and CacheManager, but it's a class by itself.
- Made the types that SettingGateway accepts for the key types dynamic (based on SettingResolver's prototype) and public throught `client.gateways.types;`.

## Removed
- Removed a bunch of extendables.
- **[PERF]** Removed `CommandMessage` proxy in favor of the brand new `KlasaMessage`. As in *Node.js 9.2.0*, Proxy creation performs 37M ops/sec, and property access 1.7M ops/sec (all pieces after Inhibitors use that proxy). As it's now removed, property access should perform around 520M ops/sec (~305 times faster).
- Removed the `messageDelete` event.
- Removed the `messageDeleteBulk` event.
- Removed the `guildCreate` event.
- Removed the `SQL` class. Replaced by GatewaySQL and Schema/Piece.
- Removed the `CacheManager` class. Replaced by CacheProviders.
- Removed `Gateway#add`. Replaced to `Schema#addKey`.
- Removed `Gateway#remove`. Replaced to `Schema#removeKey`.
- Removed `SQLProvider#serialize`, `SQLProvider#sanitize`, `SQLProvider#CONSTANTS`.

## Fixed
- A lot of things in Typings and marked a lot of private methods properly.
- Typos in languages.
- Fixed `GatewayDriver#validate` not working properly.
- The eval command being unable to catch certain errors.
- Fixed the JSON provider throwing errors.
- Fixed multiple minor issues.
- Store loader not showing error stack.
- JSON provider loading files that are not JSON.

### CODE REVIEW

**Comment by kyranet** [comment]
This PR will be continued this weekend, removing async operations from the cache.

- [ ] Remove CacheProvider support in SG, they'll use Collections, and updating them will be by reference rather than re-setting it.
- [ ] `Gateway#fetchEntry` will be removed.
- [ ] Remove async extendables for settings and i18n, they'll go back to what we had before.

Just giving SG v2 testers a heads-up about this.

Also, the `get` action for the configuration command will be fixed asap.

**Comment by kyranet** [comment]
Done a massive change, any async method is now removed. Please use `getEntry` instead. Expect more changes:

- [ ] Check <Prefix>config get.
- [ ] Check Gateway#updateArray.
- [ ] Fix schema for when the folder is empty or all keys are hidden/unconfigureable (<Prefix>config list).
- [ ] Cache updating by reference.
- [ ] Fix GatewaySQL's entry creation.

More changes will be made soon. Please be aware that the nature of a WIP branch is to change a lot/be unstable.

**Comment by kyranet** [comment]
Be careful in the next commit, testers who are using providers from pieces (RethinkDB, MySQL, SQLite...) please abstain from using them until the new providers are pushed in [klasa-pieces#11](https://github.com/dirigeants/klasa-pieces/pull/11).

Schema#force and several other methods will be rewritten to use `Provider#updateAll` between some extra methods designed to take advantage of each DB's power.

File: src/lib/settings/SchemaPiece.js [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+		const type = typeof value;
+		if (type === 'boolean' || type === 'number' || value === null) return String(value);
+		if (type === 'string') return `'${value.replace(/'/g, "''")}'`;
+		if (type === 'object') return `'${JSON.stringify(value).replace(/'/g, "''")}'`;
+		return '';
+	}
+
+}
+
+function isNumber(number) {
```
Comment (bdistin) [inline src/lib/settings/SchemaPiece.js]: Move this to Util

File: tutorials/IncludedEvents.md [hunk 2]
```diff
@@ -14,6 +14,14 @@ Replies the reason why the command was inhibited.
 
 [events/commandInhibited.js](https://github.com/dirigeants/klasa/blob/master/src/events/commandInhibited.js)
 
+## configUpdateEntry
+
+Synchronices the configs between all shards, if the bot is sharded.
```
Comment (bdistin) [inline tutorials/IncludedEvents.md]: user configs

**Comment by MrJacz** [comment]
Was the creating providers tutorial updated?

**Comment by bdistin** [comment]
No probably not

**Comment by MrJacz** [comment]
Could you do that @kyranet so I can create the MongoDB provider, thanks

**Comment by bdistin** [comment]
Look at the json rethink or mysql providers. They are up to date, just add the missing methods. They are backwards compatible

**Comment by MrJacz** [comment]
I was going to do that, guess that will do 👍

**Comment by kyranet** [comment]
Check the JSON provider, it comes with the bare basics needed for SGv2's functionality, the RethinkDB and MySQL have extra methods for extra funcionality.

**Pull request merged**